applovin_max/ios/Classes/AppLovinMAXAdViewUIComponent.m

```diff
@@ -0,0 +1,184 @@
+{
+    for (NSString *key in parameterDict)
+    {
+        NSString *value = (NSString *) parameterDict[key];
+        [self.adView setExtraParameterForKey: key value: value];
+    }
+}
+
+-(void)setLocalExtraParameters:(NSDictionary<NSString *, id> *)parameterDict
+{
+    for (NSString *key in parameterDict)
+    {
+        id value = parameterDict[key];
+        [self.adView setLocalExtraParameterForKey: key value: value];
+    }
+}
+
+- (BOOL)hasContainerView
+{
+    return self.containerView != nil;
+}
+
+- (void)attachAdView:(AppLovinMAXAdView *)view
+{
+    self.containerView = view;
+}
+
+- (void)detachAdView
+{
+    self.containerView = nil;
+    
+    [self.adView removeFromSuperview];
+}
+
+- (void)loadAd
+{
+    [self.adView loadAd];
+}
+
+- (void)destroy
+{
+    [self detachAdView];
+    
+    self.adView.delegate = nil;
+    self.adView.revenueDelegate = nil;
+}
+
+#pragma mark - MAAdDelegate Protocol
+
+- (void)didLoadAd:(MAAd *)ad
+{
+    NSDictionary *adInfo = [[AppLovinMAX shared] adInfoForAd: ad];
+    
+    [[AppLovinMAX shared] sendEventWithName:@"OnNativeUIComponentAdViewAdLoadedEvent" body: adInfo];
+    
+    if ( self.containerView )
+    {
+        [self.containerView sendEventWithName: @"OnAdViewAdLoadedEvent" body: adInfo];
+    }
+    else
+    {
+        [self setAutoRefresh: NO];
+    }
+}
+
+- (void)didFailToLoadAdForAdUnitIdentifier:(NSString *)adUnitIdentifier withError:(MAError *)error
+{
+    NSDictionary *adLoadFailedInfo = [[AppLovinMAX shared] adLoadFailedInfoForAdUnitIdentifier: adUnitIdentifier withError: error];
+    
+    [[AppLovinMAX shared] sendEventWithName:@"OnNativeUIComponentAdViewAdLoadFailedEvent" body: adLoadFailedInfo];
+    
+    if ( self.containerView )
+    {
+        [self.containerView sendEventWithName: @"OnAdViewAdLoadFailedEvent" body: adLoadFailedInfo];
+    }
+}
+
+- (void)didClickAd:(MAAd *)ad
+{
+    if ( self.containerView )
+    {
+        NSDictionary *adInfo = [[AppLovinMAX shared] adInfoForAd: ad];
+        [self.containerView sendEventWithName: @"OnAdViewAdClickedEvent" body: adInfo];
+    }
+}
+
+#pragma mark - MAAdViewAdDelegate Protocol
+
+- (void)didExpandAd:(MAAd *)ad
+{
+    if ( self.containerView )
+    {
+        NSDictionary *adInfo = [[AppLovinMAX shared] adInfoForAd: ad];
+        [self.containerView sendEventWithName: @"OnAdViewAdExpandedEvent" body: adInfo];
+    }
+}
+
+- (void)didCollapseAd:(MAAd *)ad
+{
+    if ( self.containerView )
+    {
+        NSDictionary *adInfo = [[AppLovinMAX shared] adInfoForAd: ad];
+        [self.containerView sendEventWithName: @"OnAdViewAdCollapsedEvent" body: adInfo];
+    }
+}
+
+#pragma mark - MAAdRevenueDelegate Protocol
+
+- (void)didPayRevenueForAd:(MAAd *)ad
+{
+    if ( self.containerView )
+    {
+        NSDictionary *adInfo = [[AppLovinMAX shared] adInfoForAd: ad];
+        [self.containerView sendEventWithName: @"OnAdViewAdRevenuePaidEvent" body: adInfo];
+    }
```

Comment:
Does the `didPayRevenueForAd:` callback get called when preloading an ad (but not attaching to widget tree)?

Reply:
This gets called only while it is shown.